flag_engine/identities/traits/types.py

```diff
@@ -1,13 +1,57 @@
-from typing import Union
+import re
+from decimal import Decimal
 
-from pydantic.types import confloat, constr
+from typing import Union, Any, get_args
+from typing_extensions import TypeGuard
+
+from pydantic.types import (
+    AllowInfNan,
+    StringConstraints,
+    StrictBool,
+)
+from pydantic import BeforeValidator
+from typing_extensions import Annotated
 
 from flag_engine.identities.traits.constants import TRAIT_STRING_VALUE_MAX_LENGTH
 
-TraitValue = Union[
-    None,
-    int,
-    confloat(allow_inf_nan=False),
-    bool,
-    constr(max_length=TRAIT_STRING_VALUE_MAX_LENGTH),
+_UnconstrainedTraitValue = Union[None, int, float, bool, str]
+
+
+def map_any_value_to_trait_value(value: Any) -> _UnconstrainedTraitValue:
+    if _is_trait_value(value):
+        if isinstance(value, str):
+            return _map_string_value_to_trait_value(value)
+        return value
+    if isinstance(value, Decimal):
+        if value.as_tuple().exponent:
+            return float(str(value))
+        return int(value)
+    return str(value)
```

Comment:
I think I'm right in thinking this code essentially replaces [this logic](https://github.com/Flagsmith/flagsmith-engine/pull/190/files#diff-84ff073f5f522b58426c179d8863d5f8848762db25edd9468d75f14c336f2780L15-L21), right?

I'm a little unsure about this, however, as it seems to be converting values that are passed in as strings, into e.g. integers or floats, whereas the previous logic didn't do this. I have a feeling that the API will do this anyway, so maybe it's the correct thing to do, but it is a slight change in behaviour for e.g. local evaluation mode, right?

Reply:
The previous logic did not do this because it relied on marshmallow/pydantic to perform this. With pydantic V2 the coercion logic for unions changed and I had to surface this behaviour here.